### PR TITLE
Add `voxelize_binary_mask` filter to generate `ImageData` from `PolyData`

### LINKED ISSUE #4312: Create a voxelize feature that takes (n_x, n_y, n_z) as parameters instead of just density
### Describe the feature you would like to be added.

Hi! It would be great if PyVista had a feature that allowed for the voxelization of a mesh according to a specified 3d-grid encompassing the object. The current voxelize feature allows for the adjustment of a density parameter to control the voxel size, but there is no way to directly create for example a 256 by 256 by 256 voxel grid centered on bounding box of a given mesh. For example, implementing something similar to the voxelization capabilities binvox or pyntcloud would be super useful in PyVista. It seems like perhaps this is already possible in vtk, but it's not entirely clear to me as I'm not too familiar with the inner workings of vtk. 

binvox: https://www.patrickmin.com/binvox/
pyntcloud: https://pyntcloud.readthedocs.io/en/latest/structures.html highlight=add_structure#pyntcloud.PyntCloud.add_structure

### Links to VTK Documentation, Examples, or Class Definitions.

https://vtk.org/doc/nightly/html/classvtkVoxelModeller.html

### PR DESCRIPTION
**Pull request by user27182**
### Overview

In medical image processing, it's often necessary to go from image (e.g. labelmap or binary segmentation mask) to 3D surface (e.g. using marching cubes or surface nets), do some processing, and then convert the surface back to an image. A key detail is that we want the output image to be identical in every way to the original image but with the labelmap/mask updated to reflect any changes made to the polydata surface. I.e. we want to preserve the input image extent, spacing, and orientation.

In PyVista we have `contour_labeled` to go from image to surface, but are missing methods to convert back to an image. This PR adds this new method.

TODO:
- [x] Improve API
- ~[ ] Support for multiple labels~ Will be a separate filter / PR
- [x] Better test coverage

### Details

This method is similar to `voxelize_volume` with a few key differences:
- `voxelize_volume` returns a `RectilinearGrid`; this new method returns `ImageData`
- `voxelize_volume` assumes the input image data is axis-aligned; this new method takes into account image orientation
- `voxelize_volume` only allows specifying density; this new method also allows setting `dimensions` (i.e. closes #4312)
- `voxelize_volume` uses a default spacing (density) of `mesh.length/100`; this new method uses the 10th percentile of the input's cell lengths
- `voxelize_volume` does not consider labeled input data (see #6255); ~this new method aims to support labeled inputs~ separate PR for this
- `voxelize_volume` uses `select_enclosed_points` internally which limits its used to closed surfaces only; this new method does not have this limitation
- `voxelize_volume` uses a custom implementation with `np.arange` or `np.linspace` which can cause issues with generating the correct bounds (e.g. see #6609); this new method uses vtk methods instead (using image stencils)
- The implementation is based on the "ModelToLabelMap" function from 3D Slicer, which is the gold standard (IMO) for processing medical data with VTK:
https://github.com/Slicer/Slicer/blob/7727475e9a771e6d509c994a4b1e4daa08cf0f8c/Modules/CLI/ModelToLabelMap/ModelToLabelMap.cxx

### CODE REVIEW

**Comment by GuillaumeBroggi** [comment]
I tried this new method to voxelize a mesh. As mentioned in #1743, it seems superior to the existing approaches as it does not produce stray voxels and is significantly faster.

Instead of raising an error when `reference_volume` is `None`, what about automatically generating a reference volume based on the bounds of the input mesh? I can write a commit if this seems like a good idea.

**Comment by tkoyama010** [comment]
> I tried this new method to voxelize a mesh. As mentioned in #1743, it seems superior to the existing approaches as it does not produce stray voxels and is significantly faster.
> 
> Instead of raising an error when `reference_volume` is `None`, what about automatically generating a reference volume based on the bounds of the input mesh? I can write a commit if this seems like a good idea.

@GuillaumeBroggi Thanks for your thought. You can create Pull Request for this branch if you want :)

**Comment by user27182** [comment]
That's a good suggestion for improving the API. The current error is based on the Slicer implementation that was used as an initial reference. Still a few artifacts left over from porting that over. 

The hold up for this PR at the moment is with how images are transformed in PyVista , see #6616. Once resolved I'll work on improving this PR

**Comment by user27182** [comment]
@GuillaumeBroggi The filter has been updated to no longer require a reference volume, e.g.:

``` python
from pyvista import examples
import pyvista as pv
poly = examples.load_ant()
labelmap = poly.generate_labelmap()

pl = pv.Plotter()
pl.add_mesh(labelmap.points_to_cells().threshold(0.5), color='blue')
pl.add_mesh(poly, color='lime')
pl.show()
```
Voxelized is blue, input poly data is green:
ant

Feel free to review the code and suggest changes or improvements. (Also note that it's still a draft and I may also push changes)

File: pyvista/core/filters/poly_data.py [hunk 1]
```diff
@@ -3892,3 +3897,111 @@ def protein_ribbon(self, progress_bar=False):
         alg.SetInputData(self)
         _update_alg(alg, progress_bar, "Generating Protein Ribbons")
         return _get_output(alg)
+
+    def generate_labelmap(
```
Comment (user27182) [inline pyvista/core/filters/poly_data.py:3901]: Looking for feedback on the name. The term `labelmap` is kind of domain specific to medical images, maybe something more general is better?

This method is basically the inverse surface nets from `contour_labeled`.  So we could call it something like `contour_to_labeled` to better relate the two? 

Or, with #5968, we also have the option to name _both_ filters, e.g. rename `contour_labeled` -> `image_labels_to_polydata`, and name the filter from this PR `polydata_to_image_labels`

Or, even still, we could better align the name with `voxelize` and `voxelize_volume`, and maybe call this filter `voxelize_labels`?

**Comment by user27182** [comment]
With [d4eedba](https://github.com/pyvista/pyvista/pull/6611/commits/d4eedba1b83587e559cd517d295da341b6df3f8e), the filter now dynamically sets the spacing based on the input cell length. This differs from `voxelize_volume` which uses `mesh.length/100` by default.
 
E.g. Using the same code snippet above, but with `download_bunny_coarse`:
image
And again but with full-res `download_bunny`:
image

And once again with the ant. This one is a good test because it has many large cells for the body, and tiny cells for the antennae:
image

File: pyvista/core/filters/poly_data.py (hunk 1 above)
Comment (GuillaumeBroggi) [reply pyvista/core/filters/poly_data.py:3902]: Coming from material sciences, `labelmap` is not the usual terminology, although some literature uses `labelling`.

`voxelize` is more common in this field but does not carry the idea that voxels are labeled. Another one could be `segmentation`.

**Comment by user27182** [comment]
> @user27182 I reviewed the code, and it looks great, much better than I would have done.

Thanks for the feedback!

> In my implementation, I had chosen to use `**kwargs` in case the user wants to use a custom spacing or custom origin. I think your proposition brings more functionality.

Great, so the current option to specify `spacing` or `dimensions` separately is good? I originally considered adding `origin` explicitly too, but it gets too complicated since it also needs to be rounded for the grid to match the bounds correctly. If desired, users can specify `origin` implicitly using a reference volume instead.

> `spacing_bound` is a bit misleading IMO since the code rounds `dimensions`. See #6647 for a different implementation where `rounding_func` replaces `spacing_bound` and accepts any callable. This would provide more flexibility and is less verbose.

I agree with changing `spacing_bounds`.  #6647 looks good to me.

> Also, there is a floating point error when casting the results to a `RectilinearGrid`, but I am not sure it is possible to do anything about it

It appears the error stems from the casting to rectilinear grid, perhaps this function can be adjusted (in a separate PR) to remove the errror?
https://github.com/pyvista/pyvista/blob/e1b3f1559c1a95add0f9013d48005f79fb9d3905/pyvista/core/grid.py#L845-L849

**Comment by GuillaumeBroggi** [comment]
> Great, so the current option to specify `spacing` or `dimensions` separately is good?

Yes, it works nicely like that IMO.

> It appears the error stems from the casting to rectilinear grid, perhaps this function can be adjusted (in a separate PR) to remove the error?

Thanks for pointing it out, I'll try to correct the error.

File: pyvista/core/filters/poly_data.py (hunk 1 above)
Comment (user27182) [reply pyvista/core/filters/poly_data.py:3902]: Yes I think the term `label` is perhaps a bit vague/non-obvious outside of specific fields. Maybe `voxelize_binary_mask`  is good? The method currently only works for binary masks anyway, I'm thinking a multi-label version is probably better suited for a separate filter (and separate PR), maybe call that one `voxelize_label_mask`

**Comment by user27182** [comment]
I'd say this is ready for review. Docs _might_ still fail from minor RST formatting, but otherwise is feature complete.

File: pyvista/core/filters/poly_data.py [hunk 2]
```diff
@@ -3892,3 +3898,403 @@ def protein_ribbon(self, progress_bar=False):
...
+        If no inputs are provided, the 10th ``cell_length_percentile`` is used by
+        default on systems with VTK >= 9.2. For older versions of VTK, the default
+        spacing is ``mesh_length_fraction=1/100``.
+
+        .. note::
+            For best results, ensure the input surface is a closed surface. The
+            surface is considered closed if it has zero :attr:`pyvista._PointSet.n_open_edges`.
+
+        .. note::
+            This filter returns voxels represented as point data, not :attr:`pyvista.CellTypes.VOXEL` cells..
```
Comment (user27182) [inline pyvista/core/filters/poly_data.py:3949]:
```suggestion
            This filter returns voxels represented as point data, not :attr:`pyvista.CellTypes.VOXEL` cells.
```

File: pyvista/core/filters/poly_data.py [hunk 3]
```diff
@@ -3892,3 +3898,403 @@ def protein_ribbon(self, progress_bar=False):
...
+        pyvista.ImageData
+            Generated binary mask with a ``'mask'``  point data array. The data array
+            has dtype :class:`numpy.uint8` if the foreground and background values are
+            unsigned and less than 256.
+
+        See Also
+        --------
+        pyvista.voxelize
+            Similar function that returns a :class:`pyvista.UnstructuredGrid` of
+            :attr:`pyvista.CellType.VOXEL` cells.
```
Comment (user27182) [inline pyvista/core/filters/poly_data.py:4034]:
```suggestion
            :attr:`~pyvista.CellType.VOXEL` cells.
```

File: pyvista/core/filters/poly_data.py [hunk 4]
```diff
@@ -3892,3 +3898,403 @@ def protein_ribbon(self, progress_bar=False):
...
+
+        pyvista.voxelize_volume
+            Similar function that returns a :class:`pyvista.RectilinearGrid` with cell data.
+
+        pyvista.ImageDataFilters.contour_labeled
+            Filter that generates surface contours from labeled image data. Can be
+            loosely considered as the inverse of this filter.
+
+        pyvista.ImageDataFilters.points_to_cells
+            Convert voxels represented as points to :attr:`pyvista.CellType.VOXEL`
```
Comment (user27182) [inline pyvista/core/filters/poly_data.py:4044]:
```suggestion
            Convert voxels represented as points to :attr:`~pyvista.CellType.VOXEL`
```

File: pyvista/core/filters/poly_data.py [hunk 5]
```diff
@@ -3892,3 +3898,403 @@ def protein_ribbon(self, progress_bar=False):
...
+            # Get size of poly data for computing dimensions
+            bnds = self.bounds  # type: ignore[attr-defined]
+            x_size = bnds.x_max - bnds.x_min
+            y_size = bnds.y_max - bnds.y_min
+            z_size = bnds.z_max - bnds.z_min
+            sizes = np.array((x_size, y_size, z_size))
+
+            if dimensions is None:
+                dimensions = np.array(rounding_func(sizes / initial_spacing), dtype=int)
+
```
Comment (user27182) [inline pyvista/core/filters/poly_data.py:4227]:
```suggestion
            # Get initial spacing (will be adjusted later)
            initial_spacing = _validation.validate_array3(spacing, broadcast=True)

            # Get size of poly data for computing dimensions
            bnds = self.bounds  # type: ignore[attr-defined]
            x_size = bnds.x_max - bnds.x_min
            y_size = bnds.y_max - bnds.y_min
            z_size = bnds.z_max - bnds.z_min
            sizes = np.array((x_size, y_size, z_size))

            if dimensions is None:
                dimensions = np.array(rounding_func(sizes / initial_spacing), dtype=int)

            reference_volume = pyvista.ImageData()
```

File: pyvista/core/utilities/features.py [hunk 6]
```diff
@@ -83,6 +83,18 @@ def voxelize(mesh, density=None, check_surface=True, enclosed=False, fit_bounds=
     Prior to version 0.39.0, this method improperly handled the order of
     structured coordinates.
 
+    See Also
+    --------
+    pyvista.voxelize
+        Similar function that returns a :class:`pyvista.UnstructuredGrid` of
+        :attr:`pyvista.CellType.VOXEL` cells.
```
Comment (user27182) [inline pyvista/core/utilities/features.py:90]:
```suggestion
        :attr:`~pyvista.CellType.VOXEL` cells.
```

File: pyvista/core/utilities/features.py [hunk 7]
```diff
@@ -246,6 +258,12 @@ def voxelize_volume(mesh, density=None, check_surface=True, enclosed=False, fit_
     See Also
     --------
     pyvista.voxelize
+        Similar function that returns a :class:`pyvista.UnstructuredGrid` of
+        :attr:`pyvista.CellType.VOXEL` cells.
```
Comment (user27182) [inline pyvista/core/utilities/features.py:262]:
```suggestion
        :attr:`~pyvista.CellType.VOXEL` cells.
```

**Review by GuillaumeBroggi — APPROVED** [review]
Looks good to me, that's a handy new feature.

If I am not mistaken there is a circular reference in docs, see #6651

**Review by tkoyama010 — APPROVED** [review]
LGTM

**Comment by GuillaumeBroggi** [comment]
> It appears the error stems from the casting to rectilinear grid, perhaps this function can be adjusted (in a separate PR) to remove the errror?
> 
> https://github.com/pyvista/pyvista/blob/e1b3f1559c1a95add0f9013d48005f79fb9d3905/pyvista/core/grid.py#L845-L849

Fix proposed in #6656

**Pull request merged**